pkg/activator/handler/concurrency_reporter.go

```diff
@@ -22,64 +22,85 @@ import (
 	"github.com/knative/serving/pkg/autoscaler"
 )
 
-// Channels is a structure for holding the channels for driving Stats.
-// It's just to make the NewStats signature easier to read.
-type Channels struct {
+type ConcurrencyReporter struct {
+	podName string
+
 	// Ticks with every request arrived/completed respectively
-	ReqChan chan ReqEvent
+	reqChan chan ReqEvent
 	// Ticks with every stat report request
-	ReportChan <-chan time.Time
+	reportChan <-chan time.Time
 	// Stat reporting channel
-	StatChan chan *autoscaler.StatMessage
+	statChan chan *autoscaler.StatMessage
+
+	// Contains the number of in-flight requests per-key
+	outstandingRequestsPerKey map[string]int32
+	// Contains the number of incoming requests in the current
+	// reporting period, per key.
+	incomingRequestsPerKey map[string]int32
 }
 
 // NewConcurrencyReporter instantiates a new goroutine that consumes and
 // produces from the given channels.
 // On each tick on the ReportChan, StatMessages will be sent to the
 // StatChan.
-func NewConcurrencyReporter(podName string, channels Channels) {
-
-	go func() {
-		outstandingRequestsPerKey := make(map[string]int32)
-
-		// Contains the number of incoming requests in the current
-		// reporting period, per key.
-		incomingRequestsPerKey := make(map[string]int32)
-		for {
-			select {
-			case event := <-channels.ReqChan:
-				switch event.EventType {
-				case ReqIn:
-					incomingRequestsPerKey[event.Key]++
-					outstandingRequestsPerKey[event.Key]++
-				case ReqOut:
-					outstandingRequestsPerKey[event.Key]--
+func NewConcurrencyReporter(podName string, reqChan chan ReqEvent, reportChan <-chan time.Time, statChan chan *autoscaler.StatMessage) *ConcurrencyReporter {
+	return &ConcurrencyReporter{
+		podName:    podName,
+		reqChan:    reqChan,
+		reportChan: reportChan,
+		statChan:   statChan,
+
+		outstandingRequestsPerKey: make(map[string]int32),
+		incomingRequestsPerKey:    make(map[string]int32),
+	}
+}
+
+func (cr *ConcurrencyReporter) report(now time.Time, key string, concurrency int32) {
+	requestCount := cr.incomingRequestsPerKey[key]
```

Comment:
Can we pass `requestCount` to the report method as well? That'd enable us to drop the maps from the struct and eliminate concurrent access concerns to them.